lua/lvim/updater.lua

```diff
@@ -0,0 +1,298 @@
+    if dir_exists then
+      return r
+    end
+
+    local url = "https://github.com/" .. repo .. "/archive/" .. zip_name
+
+    return r
+      :and_then(await, jobs.run({ "curl", "-LO", url }, { cwd = download_dir }))
+      :and_then(await, jobs.run({ "unzip", "-o", join_paths(download_dir, zip_name), "-d", core_install_dir }, {}))
+      :or_else(function()
+        error("download and install failed for plugin '" .. repo .. "'")
+      end)
+  end)
+end
+
+local function opendir(dir, callback)
+  uv.fs_opendir(dir, callback, 9999)
+end
+
+local async_opendir = a.wrap(opendir)
+
+-- don't want to remove any unintended directory
+local function safer_rm_rf(parent_dir, child_dir)
+  local path_to_remove = parent_dir .. path_sep .. child_dir
+  if parent_dir == nil or #parent_dir == 0 then
+    return
+  end
+  if child_dir == nil or #child_dir == 0 then
+    return
+  end
+
+  vim.fn.delete(path_to_remove, "rf")
+end
+
+local function cleanup_stale(core_install_dir, core_plugins)
+  return async(function()
+    local err, dir = await(async_opendir(core_install_dir))
+    assert(not err, "unable to open core plugin install directory")
+    local entries
+    err, entries = await(async_readdir(dir))
+    assert(not err, "unable to read core plugin install directory")
+    await(async_closedir(dir))
+
+    if entries then
+      local plugin_lookup = {}
+      for _, plug in ipairs(core_plugins) do
+        plugin_lookup[plug_extracted_dir(plug)] = true
+      end
+
+      await(a.main)
+      for _, entry in ipairs(entries) do
+        local plugin_name = entry.name
+        if plugin_name ~= nil and plugin_lookup[plugin_name] == nil then
+          safer_rm_rf(core_install_dir, plugin_name)
+        end
+      end
+    end
+
+    print("Cleaned up stale plugins in:", timer:stop(), "ms")
+    timer:start()
+  end)
+end
+
+function M:task_update(plugin_name, status)
+  print(plugin_name .. " - " .. status)
+end
+
+local packer_display_open = require("packer.display").open
+require("packer.display").open = function(self)
+  local disp = packer_display_open(self)
+  local total = 0
+  local finished = 0
+
+  local function print_update()
+    local line = "Installed " .. finished .. " of " .. total
+    vim.schedule(function()
+      vim.fn.chansend(1, { "\r" .. line .. "\n" })
+    end)
+  end
+
+  disp.task_succeeded = function()
+    finished = finished + 1
+    print_update()
+  end
+  disp.task_failed = function(_, plugin_name)
+    print("!! " .. plugin_name .. ": failed to install!")
+    finished = finished + 1
+    print_update()
+  end
+  disp.task_start = function()
+    require("packer.display").status.running = true
+    total = total + 1
+    print_update()
+  end
+  return disp
+end
+
+function M.init(opts)
+  local core_install_dir = opts.core_install_dir or vim.fn.stdpath "data" .. "/core"
+  assert(vim.fn.executable "curl", "curl is not installed")
+  assert(vim.fn.executable "unzip", "unzip is not installed")
```

Comment:
These should be moved into `download_and_install()` or wrapped to not execute in `LVIM_DEV`

Reply:
Good catch!